### PR TITLE
Optimize Append.ToList() & MoveNext() performance

### PR DESCRIPTION
**Pull request by timandy**

### CODE REVIEW

File: src/System.Linq/src/System/Linq/AppendPrepend.SpeedOpt.cs [hunk 1]
```diff
@@ -189,10 +189,12 @@ public override List<TSource> ToList()
                 list.AddRange(_source);
                 if (_appended != null)
                 {
-                    IEnumerator<TSource> e = _appended.GetEnumerator(_appendCount);
-                    while (e.MoveNext())
+                    using (IEnumerator<TSource> e = _appended.GetEnumerator(_appendCount))
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/AppendPrepend.SpeedOpt.cs]: This isn't necessary; _appended here is a SingleLinkedNode, whose `GetEnumerator(int)` method calls `ToArray(int)` on it and then returns that array's enumerator.  That enumerator's Dispose is a nop.

But this highlights that there's unnecessary overhead here, as we don't actually need the enumerator allocation or the resulting interface calls at all.  Instead, we can make:
https://github.com/dotnet/corefx/blob/7dce7868e3b3023ee96edebb96042ca01a8ec368/src/System.Linq/src/System/Linq/SingleLinkedNode.cs#L101
be `public`, and then change this code from:
```C#
IEnumerator<TSource> e = _appended.GetEnumerator(_appendCount);
while (e.MoveNext()) list.Add(e.Current);
```
to instead be:
```
list.AddRange(_appended.ToArray(_appendCount));
```
This should be strictly better than what's currently there.

Comment (timandy) [reply src/System.Linq/src/System/Linq/AppendPrepend.SpeedOpt.cs]: @stephentoub  Got it.
May be rewrite SingleLinkedNode.GetEnumerator like below is better, Am I right?
```C#
        public IEnumerator<TSource> GetEnumerator(int count)
        {
            Debug.Assert(count == GetCount());

            for (SingleLinkedNode<TSource> node = this; node != null; node = node.Linked)
            {
                yield return node.Item;
            }
        }
```

Comment (stephentoub) [reply src/System.Linq/src/System/Linq/AppendPrepend.SpeedOpt.cs]: > May be rewrite SingleLinkedNode.GetEnumerator like below is better, Am I right?

No, that would result in the elements being stored in the wrong order.

Comment (timandy) [reply src/System.Linq/src/System/Linq/AppendPrepend.SpeedOpt.cs]: You are right. I forgot SingleLinkedNode is a reversed linked list.Closed

**Pull request closed without merging**